From f81cd6cdff1213f90de8f012489017510e3d6ff4 Mon Sep 17 00:00:00 2001
From: William Vicary <will@3whitehats.com>
Date: Tue, 18 Feb 2020 19:45:30 +0000
Subject: [PATCH] Handle exceptions during response creation

Catches exceptions thrown during the createResponse method of the easy handler to avoid crashes during the headers method.
---
 src/Handler/CurlFactory.php | 19 ++++++++++++++++++-
 src/Handler/EasyHandle.php  |  3 +++
 2 files changed, 21 insertions(+), 1 deletion(-)

diff --git a/src/Handler/CurlFactory.php b/src/Handler/CurlFactory.php
index 4a28a96eb..0c85c53fc 100644
--- a/src/Handler/CurlFactory.php
+++ b/src/Handler/CurlFactory.php
@@ -165,6 +165,18 @@ private static function createRejection(EasyHandle $easy, array $ctx)
             CURLE_GOT_NOTHING          => true,
         ];
 
+        if ($easy->createResponseException) {
+            return \GuzzleHttp\Promise\rejection_for(
+                new RequestException(
+                    'An error was encountered while creating the response',
+                    $easy->request,
+                    $easy->response,
+                    $easy->createResponseException,
+                    $ctx
+                )
+            );
+        }
+
         // If an exception was encountered during the onHeaders event, then
         // return a rejected promise that wraps that exception.
         if ($easy->onHeadersException) {
@@ -562,7 +574,12 @@ private function createHeaderFn(EasyHandle $easy)
             $value = trim($h);
             if ($value === '') {
                 $startingResponse = true;
-                $easy->createResponse();
+                try {
+                    $easy->createResponse();
+                } catch (\Exception $e) {
+                    $easy->createResponseException = $e;
+                    return -1;
+                }
                 if ($onHeaders !== null) {
                     try {
                         $onHeaders($easy->response);
diff --git a/src/Handler/EasyHandle.php b/src/Handler/EasyHandle.php
index 7754e9111..b59008d6a 100644
--- a/src/Handler/EasyHandle.php
+++ b/src/Handler/EasyHandle.php
@@ -37,6 +37,9 @@ final class EasyHandle
     /** @var \Exception Exception during on_headers (if any) */
     public $onHeadersException;
 
+    /** @var \Exception Exception during createResponse (if any) */
+    public $createResponseException;
+
     /**
      * Attach a response to the easy handle based on the received headers.
      *
